### ISSUE TITLE
clarify --registry-mirror gotcha statement

### ISSUE THREAD

**Issue by peterlynch** [docker/distribution#1483]
https://github.com/docker/distribution/blame/d4910831a339fa96a21a4e473312c2155e2570d9/docs/mirror.md#L21

States the following about the --registry-mirror option:

> Gotcha
> It’s currently not possible to mirror another private registry. Only the central Hub can be mirrored.

The statement is vague enough that I need clarification on the following questions:

1. Is this referring to **any ** private registry implementation or just Docker's private registry on-premise product?
2. Does this mean that special handling is required on the registry side (server) in order to function as a argument for --registry-mirror? Is there some check performed by the daemon that enforces this option to only work with Docker Hub? If there is some special handling required outside of the typical Docker API, what is it?

**Comment by dmp42** [comment]
1. this is referring to anything that is not the Hub - any third-party registry, no matter what the implementation is
2. no special handling is required on the registry side AFAIK (though someone may correct me on this) - the current limitation is in the docker engine

Though an often requested feature, the reason we haven't allowed this yet is that in the current state of affairs it would lead to namespace collisions.

We also noticed that many people request this feature while they don't understand the design of the registry that already let them achieve what they want (eg: leveraging the backend, or a middleware for delivery, or even the plain "http cacheability" of the registry).

If you are interested in this, this here is for you: #1431 (starting with collecting use cases).

**Comment by davidkarlsen** [comment]
So what you are saying @dmp42 is that --registry-mirror will only work to mirror dockerhub regardless of what server-implementation is serving that mirror, and that it's a docker-engine (not mirror) limitation.
In that case https://github.com/docker/docker/issues/20097 is not a valid usecase and can be closed - but one should note that this actually worked with engine versions < 1.10.

**Comment by dmp42** [comment]
@davidkarlsen I don't think we ever supported mirroring private registries with the v2 protocol (cc @RichardScothern for confirmation). The documentation that @peterlynch linked is from August (before 1.9 was released) and further confirm this.

I haven't read docker/docker#20097 which is quite long, though @aaronlehmann  did and can provide additional info about this.

Nevertheless, I strongly encourage you to provide use-cases and feedback on #1431

**Comment by davidkarlsen** [comment]
@dmp42 It did not mirror private repositories per se. The registry behind our --registry-mirror host:port is a normal mirror that federates several repositories (among them a local one, not hub-private one). E.g. its a logical mirror.

Example:
hostLogicalRepo:port is what we put in the --registry-mirror option
hostLogicalRepo:port federates docker.io:443 and ourOwnRegistry:port behind the scenes and looks like one logical registry as far as docker-engine as a client goes.

Pulling hostLogicalRepo:port/elasticsearch:2.2 would lookup docker.io/elasticsearch:2.2 and serve me that

Pulling hostLogicalRepo:port/myInternalImg:latest would lookup docker.io/myInternalImg:latest, not find it, lookup ourOwnRegistry:port/myInternalImg:latest and serve me that - or rather that used to work  with engine <1.10 - no longer on 1.10 - that's the essence of #20097 .

**Comment by dmp42** [comment]
@davidkarlsen I understand.

This scenario was never supported. Furthermore, this usage is breaking one of the fundamental expectation and core feature of images namespace: guarantee of unicity and reproducibility. By effectively hijacking the global namespace here, you are creating a breakage in the ecosystem, where a given name ("myInternalImg") may refer to different bits for different users.

This is not purely me being anal over it - doing this will never work with Content Trust enabled, since you are not entitled to sign content under the docker.io/ namespace.

Long story short:

 * I really advise you to walk away from that approach to "merge" namespaces and registries under the same space, because of all the above
 * though: I'm happy to discuss the actual "user scenario" you want to solve, and figure out better ways to achieve what you want - what problem is this trying to solve? etc

We are usually available on irc (#docker-distribution on freenode) or on the google group if you want.

Hope that helps.

**Comment by dmp42** [comment]
@aaronlehmann I'm still curious to understand why @davidkarlsen scenario worked previously and does no more.

**Comment by aaronlehmann** [comment]
@dmp42: I am also curious about this. We had quite a bit of discussion about this in https://github.com/docker/docker/issues/20097, and despite @davidkarlsen being extremely helpful in answering repeated rounds of questions, I never managed to get a similar setup to work with Docker 1.9 and Registry v2. Docker There may be something nonstandard in Sonatype's registry to support credential handling for this federated setup.

**Comment by clemenko** [comment]
There is a VERY valid use case of isolated networks that want to setup caching mirrors for routed vlans for segregated clusters/networks. We run several hadoop cluster that don't have complete access to our primary lan. The entire namespace is handled by our primary private registry.

**Comment by hulu1522** [comment]
I too would like to see a docker-registry implementation that acts as a pure cache server.  If nothing else add an option to enable the registry as a "cache-only" server with it pointing at another host.  That other host will then do all the proxy / mirror options.   The cache registry will literally just forward pull requests on and cache them locally to reduce the network traffic and increase local pull speeds.  It doesn't need any other special configs (Storage, Type="cache", Registry=someV2.registry).

**Comment by CMYanko** [comment]
Near as I can tell the client/engine will only use docker hub creds for a registry mirror. If I make a local user to mimic that account, I can get the mirror to work in my local environment.
